include/zephyr/devicetree.h

```diff
@@ -3787,6 +3787,70 @@
  */
 #define DT_ON_BUS(node_id, bus) IS_ENABLED(DT_CAT3(node_id, _BUS_, bus))
 
+/**
+ * @}
+ */
+
+/**
+ * @defgroup devicetree-generic-endpoint Endpoint helpers
+ * @ingroup devicetree
+ * @{
+ */
+
+/**
+ * @brief Get the remote endpoint node connected to a local endpoint.
+ *
+ * Devices can be interconnected through ports and endpoints.
+ *
+ * Example devicetree overlay:
+ *
+ * @code{.dts}
+ *	&source {
+ *		port {
+ *			source_out: endpoint {
+ *				remote-endpoint-label = "sink_in";
+ *				source-property = <1>;
+ *			};
+ *		};
+ *	};
+ *
+ *	&sink {
+ *		port {
+ *			sink_in: endpoint {
+ *				remote-endpoint-label = "source_out";
+ *				sink-property = <2>;
+ *			};
+ *		};
+ *	};
+ * @endcode
+ *
+ * @c DT_REMOTE_ENDPOINT(endpoint_node) permits to access the other endpoint
+ * connected to @c endpoint_node.
+ *
+ * Example usage: starting from the sink, access the source endpoint connected
+ * to the @c sink_in endpoint:
+ *
+ * @code{.c}
+ *	DT_PROP(DT_REMOTE_ENDPOINT(DT_NODELABEL(sink_in)), source_property)
+ * @endcode
+ *
+ * Example usage, starting from the source, to access the sink endpoint
+ * connected to the @c source_out endpoint:
+ *
+ * @code{.c}
+ *	DT_PROP(DT_REMOTE_ENDPOINT(DT_NODELABEL(source_out)), sink_property)
+ * @endcode
+ *
+ * @note Once circular phandle references are supported in Zephyr devicetree,
+ *       @c remote-endpoint-label (string) may be changed into
+ *       @c remote-endpoint (phandle).
+ *
+ * @param node The local endpoint node to use.
+ *
+ * @return The remote endpoint node connected to @p node.
+ */
+#define DT_REMOTE_ENDPOINT(node) DT_NODELABEL(DT_STRING_TOKEN(node, remote_endpoint_label))
```

Comment:
Given real macro usages in drivers as below:

Case 1:
```
#define DEVICE_DT_INST_GET_SOURCE_DEV(n)                                                           \
	DEVICE_DT_GET(DT_PARENT(DT_GPARENT(DT_NODELABEL(DT_STRING_TOKEN(                           \
		DT_CHILD(DT_INST_CHILD(n, port), endpoint), remote_endpoint_label)))))
```
the node parameter in this case will be `DT_CHILD(DT_INST_CHILD(n, port), endpoint)`

Case 2:

```
#define DEVICE_DT_INST_GET_SENSOR_DEV(n)                                                           \
	DEVICE_DT_GET(DT_GPARENT(DT_NODELABEL(                                                     \
		DT_STRING_TOKEN(DT_CHILD(DT_CHILD(DT_INST_CHILD(n, ports), port_1), endpoint),     \
				remote_endpoint_label))))
```

the node parameter will be `DT_CHILD(DT_CHILD(DT_INST_CHILD(n, ports), port_1), endpoint)`

There's still `endpoint`common for all cases. Is it possible to take endpoint in the DT_REMOTE_ENDPOINT api ?

Also, to retrieve the peer device object, we always has at least `DT_GPARENT`, is it possible to also take this in the macro API ?

Reply:
It is possible to reduce the overhead by putting endpoint in common to each, but then comes looping over every endpoint:

```c
#define ENDPOINT(n) DT_PROP(DT_REMOTE_ENDPOINT(n), link_parameter),

int link_parameters[] = {
        DT_FOREACH_CHILD(DT_INST_CHILD(DT_DRV_INST(0), port), ENDPOINT)
};
```

So this means that it is also useful to expose an intermediate macro that can be used in both case, only wrapping the remote-endpoint-label -> nodelabel -> node conversion.

It will also be possible to have some macro using `COND_CODE_1()` to make the "ports" vs "port" abstraction.

```c
#define DT_PORTS(n) \
COND_CODE_1(DT_HAS_PROP(n, ports), (DT_CHILD(n, ports)), (n))

#define DT_INST_PORTS(n) \
DT_PORTS(DT_DRV_INST(n))

/* Example */
DEVICE_DT_GET(DT_GPARENT(DT_REMOTE_ENDPOINT(
        DT_CHILD(DT_CHILD(DT_INST_PORTS(0), port_3, endpoint_2)))))
```

This will return the node that contains all the ports, for instance:

```c
#define DT_PORTS_REMOTE_ENDPOINT(n, port, endpoint) \
DT_REMOTE_ENDPOINT(DT_CHILD(DT_CHILD(DT_PORTS(DT_DRV_INST(0)), port), endpoint))

#define DT_INST_PORTS_REMOTE_ENDPOINT(n, port, endpoint) \
DT_PORTS(DT_DRV_INST(n))

/* Example */
DEVICE_DT_GET(DT_GPARENT(DT_INST_PORTS_REMOTE_ENDPOINT(0, port_3, endpoint_1)))
```

Then, there is going *down* from the other device, and we might need the same check for "ports" present or not.

```c
#define DT_ENDPOINT_DEVICE(n) \
COND_CODE_1(DT_PROP_HAS(DT_GPARENT(n), port), (DT_GPARENT(n)), (DT_PARENT(DT_GPARENT(n))))

/* Example */
DEVICE_DT_GET(DT_ENDPOINT_DEVICE(DT_INST_PORTS_REMOTE_ENDPOINT(0, port_3, endpoint_1)))
```

And then why not combine them two, as the most frequent use-case for simple drivers would be to pick the device connected through a particular endpoint:

```c
#define DT_REMOTE_DEVICE(n, port, endpoint) \
DT_ENDPOINT_DEVICE(DT_INST_PORTS_REMOTE_ENDPOINT(0, port, endpoint))

#define DT_INST_REMOTE_DEVICE(n, port, endpoint) \
DT_REMOTE_DEVICE(DT_DRV_INST(n), port, endpoint)

/* Example */
DEVICE_DT_GET(DT_INST_REMOTE_DEVICE(0, port, endpoint))
```

Would it make sense to decompose the implementation this much?

The intermediate macros would need to be implemented anyway to avoid making them too large, so we might as well document and test them individually.